sys/usb/usbus/dfu/dfu.c

```diff
@@ -0,0 +1,261 @@
+    /* reset offset value for next transfer */
+    offset = 0;
+    /* Erase page content */
+    flashpage_write(page_cpt, NULL);
+    /* Write buffer into flash */
+    int ret = flashpage_write_and_verify(page_cpt, page_mem);
+    page_cpt++;
+    if (ret != FLASHPAGE_OK) {
+         /* TODO: Set DFU status variable accordingly */
+         return;
+    }
+    /* Clear buffer for next transfer */
+    memset(page_mem, 0, sizeof(page_mem));
+}
+void usbus_dfu_init(usbus_t *usbus, usbus_dfu_device_t *handler, unsigned mode)
+{
+    DEBUG("DFU: initialization\n");
+    assert(usbus);
+    assert(handler);
+    memset(handler, 0, sizeof(usbus_dfu_device_t));
+    handler->usbus = usbus;
+    handler->handler_ctrl.driver = &dfu_driver;
+    handler->mode = mode;
+    dfu_state = (handler->mode) ? dfu_idle : app_idle;
+    usbus_register_event_handler(usbus, (usbus_handler_t *)handler);
+}
+
+static void _init(usbus_t *usbus, usbus_handler_t *handler)
+{
+    usbus_dfu_device_t *dfu = (usbus_dfu_device_t*) handler;
+
+    /* Set up descriptor generators */
+    dfu->dfu_descr.next = NULL;
+    dfu->dfu_descr.funcs = &_dfu_descriptor;
+    dfu->dfu_descr.arg = dfu;
+
+    /* Configure Interface 0 as control interface */
+    dfu->iface.class = USB_DFU_INTERFACE;
+    dfu->iface.subclass = USB_DFU_SUBCLASS_DFU;
+
+    dfu->iface.protocol = ((dfu->mode) ? USB_DFU_PROTOCOL_DFU_MODE :
+                                USB_DFU_PROTOCOL_RUNTIME_MODE);
+
+    dfu->iface.descr_gen = &dfu->dfu_descr;
+    dfu->iface.handler = handler;
+    dfu->iface.alts = NULL;
+
+    /* Add interfaces to the stack */
+    usbus_add_interface(usbus, &dfu->iface);
+    usbus_handler_set_flag(handler, USBUS_HANDLER_FLAG_RESET);
+}
+
+static int _control_handler(usbus_t *usbus, usbus_handler_t *handler,
+                          usbus_control_request_state_t state,
+                          usb_setup_t *setup)
+{
+    (void)usbus;
+    (void)state;
+    (void)handler;
+    (void)setup;
+
+    return 1;
+}
+
+
+static void _transfer_handler(usbus_t *usbus, usbus_handler_t *handler,
+                             usbdev_ep_t *ep, usbus_event_transfer_t event)
+{
+    (void)event;
+    (void)usbus;
+    (void)handler;
+    (void)ep;
+}
+
+void dfu_control_req(usbus_t *usbus, usbus_control_handler_t *handler)
+{
+    usb_setup_t *pkt = &handler->setup;
+    usbopt_enable_t enable = USBOPT_DISABLE;
+
+    DEBUG("DFU control request:%d\n", pkt->request);
+    switch(pkt->request) {
+        case DFU_DETACH:
+            /* Detach USB bus */
+            usbdev_set(usbus->dev, USBOPT_ATTACH, &enable, sizeof(usbopt_enable_t));
+            /* Start a new usbus instance with a single interface */
+            /* TODO: properly integrate this part to riotboot */
+            usbus_init(&usbus2, usbdev_get_ctx(0));
+            usbus_dfu_init(&usbus2, &dfu2, USBUS_DFU_MODE);
+            usbus_create(_stack, USBUS_STACKSIZE, USBUS_PRIO, USBUS_TNAME, &usbus2);
```

Comment:
How would this normally work? reboot to bootloader?

Reply:
Answer in next PR 
-> [here](https://github.com/dylad/RIOT/blob/ebcc608a474ed5d170cb3a06710fa6a49ba68baa/sys/usb/usbus/dfu/dfu.c#L184)

Reply:
How about marking this experimental in the mean time then

Reply:
I also added a line to the comment.